File: android/word/src/main/java/com/dreampany/word/data/source/firestore/FirestoreWordDataSource.java
/*
package com.dreampany.word.data.source.firestore;

import android.graphics.Bitmap;

import com.dreampany.firebase.RxFirebaseFirestore;
import com.dreampany.network.manager.NetworkManager;
import com.dreampany.word.data.model.Word;
import com.dreampany.word.data.source.api.WordDataSource;
import com.dreampany.word.misc.Constants;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Singleton;

import io.reactivex.Maybe;

*/
/**
 * Created by Hawladar Roman on 9/3/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 *//*

@Singleton
public class FirestoreWordDataSource implements WordDataSource {

    private static final String WORDS = Constants.FirebaseKey.WORDS;

    private final NetworkManager network;
    private final RxFirebaseFirestore firestore;

    public FirestoreWordDataSource(NetworkManager network,
                                   RxFirebaseFirestore firestore) {
        this.network = network;
        this.firestore = firestore;
    }

    @Override
    public Word getTodayItem() {
        return null;
    }

    @Override
    public Maybe<Word> getTodayItemRx() {
        return null;
    }

    @Override
    public boolean isEmpty() {
        return false;
    }

    @Override
    public Maybe<Boolean> isEmptyRx() {
        return null;
    }

    @Override
    public int getCount() {
        return 0;
    }

    @Override
    public Maybe<Integer> getCountRx() {
        return null;
    }

    @Override
    public boolean isExists(Word word) {
        return false;
    }

    @Override
    public Maybe<Boolean> isExistsRx(Word word) {
        return null;
    }

    @Override
    public long putItem(Word word) {
       Throwable error = firestore.setItemRx(WORDS, word.getId(), word).blockingGet();
       if (error == null) {
           return 0;
       }
       return -1;
    }

    @Override
    public Maybe<Long> putItemRx(Word word) {
        //return firestore.setItemRx(WORDS, word.getWord(), word).toMaybe();
        return null;
    }

    @Override
    public List<Long> putItems(List<? extends Word> words) {
        return null;
    }

    @Override
    public Maybe<List<Long>> putItemsRx(List<? extends Word> words) {
        return null;
    }

    @Override
    public int delete(Word word) {
        return 0;
    }

    @Override
    public Maybe<Integer> deleteRx(Word word) {
        return null;
    }

    @Override
    public Word getItem(String id) {
        return null;
    }

    @Override
    public Maybe<Word> getItemRx(String id) {
        return null;
    }

    @Override
    public Word getItem(String word, boolean full) {
        Word result = getItemRx(word, full).blockingGet();
        return result;
    }

    @Override
    public Maybe<Word> getItemRx(String word, boolean full) {
        return firestore.getItemRx(WORDS, word, Word.class);
    }

    @Override
    public List<Word> getItems(List<String> ids) {
        return null;
    }

    @Override
    public Maybe<List<Word>> getItemsRx(List<String> ids) {
        return null;
    }

    @Override
    public List<Word> getSearchItems(String query, int limit) {
        return null;
    }

    @Override
    public List<Word> getCommonItems() {
        return null;
    }

    @Override
    public List<Word> getAlphaItems() {
        return null;
    }

    @Override
    public Maybe<List<Word>> getItemsRx(Bitmap bitmap) {
        return null;
    }

    @Override
    public List<String> getRawWords() {
        return null;
    }

    @Override
    public Maybe<List<String>> getRawWordsRx() {
        return null;
    }

    @NotNull
    @Override
    public List<Long> delete(@NotNull List<? extends Word> words) {
        return null;
    }

    @NotNull
    @Override
    public Maybe<List<Long>> deleteRx(@NotNull List<? extends Word> words) {
        return null;
    }

    @NotNull
    @Override
    public List<Word> getItems() {
        return null;
    }

    @NotNull
    @Override
    public Maybe<List<Word>> getItemsRx() {
        return null;
    }

    @NotNull
    @Override
    public ArrayList<Word> getItems(int limit) {
        return null;
    }

    @NotNull
    @Override
    public Maybe<List<Word>> getItemsRx(int limit) {
        return null;
    }
}
*/

File: android/word/src/main/kotlin/com/dreampany/word/data/source/firestore/FirestoreWordDataSource.kt
package com.dreampany.word.data.source.firestore

import android.graphics.Bitmap
import com.dreampany.firebase.RxFirebaseFirestore
import com.dreampany.network.manager.NetworkManager
import com.dreampany.word.data.model.Word
import com.dreampany.word.data.source.api.WordDataSource
import com.dreampany.word.misc.Constants
import io.reactivex.Maybe
import javax.inject.Singleton

/**
 * Created by roman on 2019-07-13
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class FirestoreWordDataSource(
    val network: NetworkManager,
    val firestore: RxFirebaseFirestore
) : WordDataSource {
    override fun isExists(word: String): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getTodayItem(): Word {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getTodayItemRx(): Maybe<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItem(word: String, full: Boolean): Word? {
        return getItemRx(word, full).blockingGet()
    }

    override fun deleteRx(t: Word): Maybe<Int> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItem(id: String): Word? {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemRx(word: String, full: Boolean): Maybe<Word> {
        return firestore.getItemRx(WORDS, word, Word::class.java)
    }

    override fun getItemRx(id: String): Maybe<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItems(ids: List<String>): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItems(): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItems(limit: Int): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(ids: List<String>): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(bitmap: Bitmap): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(limit: Int): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getSearchItems(query: String, limit: Int): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getCommonItems(): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getAlphaItems(): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getRawWords(): List<String> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getRawWordsRx(): Maybe<List<String>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun isEmpty(): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun isEmptyRx(): Maybe<Boolean> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getCount(): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getCountRx(): Maybe<Int> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun isExists(t: Word): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun isExistsRx(t: Word): Maybe<Boolean> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun putItem(t: Word): Long {
        val error = firestore.setItemRx<Word>(WORDS, t.id, t).blockingGet()
        return if (error == null) { 0 } else -1
    }

    override fun putItemRx(t: Word): Maybe<Long> {
        return Maybe.create { emitter ->
            val result = putItem(t)
            if (emitter.isDisposed) {
                return@create
            }
            emitter.onSuccess(result)
        }
    }

    override fun putItems(ts: List<Word>): List<Long> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun putItemsRx(ts: List<Word>): Maybe<List<Long>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun delete(t: Word): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun delete(ts: List<Word>): List<Long> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun deleteRx(ts: List<Word>): Maybe<List<Long>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    private val WORDS = Constants.FirebaseKey.WORDS
}